### PR TITLE
expose the custom endpoints from RouterServiceFactory

### LINKED ISSUE #1400: Prometheus endpoint 404-ing when enabled
**Describe the bug**
The Prometheus endpoint is 404ing, even when Prometheus is enabled and when I've sent requests to the Router.

**To Reproduce**
1. Checkout current `main`: 671c413cda687438cd251da24eb9f65532b60584
1. In one shell, run the Router with [Prometheus enabled per the documentation](https://www.apollographql.com/docs/router/configuration/metrics/#using-prometheus):

   > This creates the https://supergraph.demo.starstuff.dev/ supergraph and a really simple YAML file via two temp files and passes them into the Router.

   ```shell
   config="$(mktemp)" &&
     echo "telemetry:\n metrics:\n  prometheus:\n   enabled: true" > "$config" &&
       graph="$(mktemp)" &&
         curl -L https://supergraph.demo.starstuff.dev/ -o "$graph" &&
           cargo run -- --supergraph "$graph" --config "$config"
   ```

3. In another shell, send at least one query so _something_ would be expected to be emitted.  It's worth noting that the page at http://127.0.0.1:4000/plugins/apollo.telemetry/prometheus is NOT BLANK, but instead returning a _404 Not Found_.  Anyhow, send the query:

    ```shell
    curl --request POST \
        --header 'content-type: application/json' \
        --url http://localhost:4000/ \
        --data '{"query":"{\n  topProducts {\n    inStock\n  }\n}"}'
    ```

4. Behold the http://localhost:4000/plugins/apollo.telemetry/prometheus page returns a 404.

**Expected behavior**
I'd expect something along the lines of what the documentation [suggests](https://www.apollographql.com/docs/router/configuration/metrics/#using-prometheus) should appear:

  ```
  # HELP http_request_duration_seconds Total number of HTTP requests made.
  # TYPE http_request_duration_seconds histogram
  http_request_duration_seconds_bucket{le="0.5"} 1
  http_request_duration_seconds_bucket{le="0.9"} 1
  ---SNIP---
  ```

**Output**

<details>
<summary>Lots of console output</summary>
<pre>
     Running `target/debug/router --supergraph /tmp/tmp.vxhVvuR8uS --config /tmp/tmp.38yepkOWca --log debug`
2022-07-14T21:31:45.116357Z  INFO apollo_router::executable: Apollo Router v0.11.0 // (c) Apollo Graph, Inc. // Licensed as ELv2 (https://go.apollo.dev/elv2)
2022-07-14T21:31:45.359590Z DEBUG apollo_router::state_machine: starting
2022-07-14T21:31:45.360063Z  INFO apollo_router::state_machine: transitioned to startup
2022-07-14T21:31:45.360189Z DEBUG apollo_router::state_machine: starting http
2022-07-14T21:31:45.409472Z DEBUG apollo_router::configuration: adding plugin apollo.telemetry with user provided configuration
2022-07-14T21:31:45.409620Z DEBUG apollo_router::router_factory: creating plugin: 'apollo.telemetry' with configuration:
{
  "metrics": {
    "prometheus": {
      "enabled": true
    }
  }
}
2022-07-14T21:31:45.410210Z DEBUG apollo_router::plugins::telemetry::metrics::apollo: configuring Apollo metrics
2022-07-14T21:31:45.411480Z  INFO apollo_router::router_factory: list of plugins plugin_details=[("apollo.telemetry", "apollo_router::plugins::telemetry::Telemetry"), ("experimental.include_subgraph_errors", "apollo_router::plugins::include_subgraph_errors::IncludeSubgraphErrors"), ("apollo.csrf", "apollo_router::plugins::csrf::Csrf")]
2022-07-14T21:31:45.411617Z DEBUG apollo_router::router_factory: activating plugin apollo_router::plugins::telemetry::Telemetry
2022-07-14T21:31:45.412284Z DEBUG apollo_router::router_factory: activated plugin apollo_router::plugins::telemetry::Telemetry
2022-07-14T21:31:45.412338Z DEBUG apollo_router::router_factory: activating plugin apollo_router::plugins::include_subgraph_errors::IncludeSubgraphErrors
2022-07-14T21:31:45.412355Z DEBUG apollo_router::router_factory: activated plugin apollo_router::plugins::include_subgraph_errors::IncludeSubgraphErrors
2022-07-14T21:31:45.412378Z DEBUG apollo_router::router_factory: activating plugin apollo_router::plugins::csrf::Csrf
2022-07-14T21:31:45.412405Z DEBUG apollo_router::router_factory: activated plugin apollo_router::plugins::csrf::Csrf
2022-07-14T21:31:45.744216Z  INFO apollo_router::axum_http_server_factory: GraphQL endpoint exposed at http://127.0.0.1:4000/ 🚀
2022-07-14T21:31:45.744353Z  INFO apollo_router::state_machine: transitioned to running
2022-07-14T21:31:45.744455Z DEBUG apollo_router::state_machine: ignoring message transition NoMoreConfiguration
2022-07-14T21:31:45.744486Z  INFO apollo_router::state_machine: transitioned to running
2022-07-14T21:31:45.744527Z DEBUG apollo_router::state_machine: ignoring message transition NoMoreSchema
2022-07-14T21:31:45.744545Z  INFO apollo_router::state_machine: transitioned to running
2022-07-14T21:32:02.507098Z DEBUG hyper::proto::h1::io: parsed 5 headers
2022-07-14T21:32:02.507189Z DEBUG hyper::proto::h1::conn: incoming body is content-length (51 bytes)
2022-07-14T21:32:02.507477Z DEBUG hyper::proto::h1::conn: incoming body completed
2022-07-14T21:32:02.508356Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}: tower_http::trace::on_request: started processing request
2022-07-14T21:32:02.510234Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}: tower::buffer::worker: service.ready=true processing request
2022-07-14T21:32:02.547353Z DEBUG hyper::client::connect::dns: resolving host="products.demo.starstuff.dev"
2022-07-14T21:32:02.549408Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: hyper::client::connect::http: connecting to [2a09:8280:1::1018]:443
2022-07-14T21:32:02.553688Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: hyper::client::connect::http: connected to [2a09:8280:1::1018]:443
2022-07-14T21:32:02.562374Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: h2::client: binding client connection
2022-07-14T21:32:02.562689Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: h2::client: client connection bound
2022-07-14T21:32:02.562883Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: h2::codec::framed_write: send frame=Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
2022-07-14T21:32:02.563994Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=WindowUpdate { stream_id: StreamId(0), size_increment: 5177345 }
2022-07-14T21:32:02.564138Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: hyper::client::pool: pooling idle connection for ("https", products.demo.starstuff.dev)
2022-07-14T21:32:02.564994Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Settings { flags: (0x0), initial_window_size: 1048576, max_frame_size: 16384, max_header_list_size: 16777216 }
2022-07-14T21:32:02.565257Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=Settings { flags: (0x1: ACK) }
2022-07-14T21:32:02.565606Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=WindowUpdate { stream_id: StreamId(0), size_increment: 983041 }
2022-07-14T21:32:02.565944Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=Headers { stream_id: StreamId(1), flags: (0x4: END_HEADERS) }
2022-07-14T21:32:02.566396Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=Data { stream_id: StreamId(1), flags: (0x1: END_STREAM) }
2022-07-14T21:32:02.567142Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Settings { flags: (0x1: ACK) }
2022-07-14T21:32:02.567374Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::proto::settings: received settings ACK; applying Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
2022-07-14T21:32:02.869605Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Headers { stream_id: StreamId(1), flags: (0x4: END_HEADERS) }
2022-07-14T21:32:02.870250Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(1) }
2022-07-14T21:32:02.870483Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="products" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=products.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(1), flags: (0x1: END_STREAM) }
2022-07-14T21:32:02.877151Z DEBUG hyper::client::connect::dns: resolving host="inventory.demo.starstuff.dev"
2022-07-14T21:32:02.878398Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: hyper::client::connect::http: connecting to [2a09:8280:1::6:1068]:443
2022-07-14T21:32:02.881358Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: hyper::client::connect::http: connected to [2a09:8280:1::6:1068]:443
2022-07-14T21:32:02.886933Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: h2::client: binding client connection
2022-07-14T21:32:02.887328Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: h2::client: client connection bound
2022-07-14T21:32:02.887612Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: h2::codec::framed_write: send frame=Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
2022-07-14T21:32:02.888872Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Settings { flags: (0x0), initial_window_size: 1048576, max_frame_size: 16384, max_header_list_size: 16777216 }
2022-07-14T21:32:02.889079Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}: hyper::client::pool: pooling idle connection for ("https", inventory.demo.starstuff.dev)
2022-07-14T21:32:02.889116Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=Settings { flags: (0x1: ACK) }
2022-07-14T21:32:02.889333Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=WindowUpdate { stream_id: StreamId(0), size_increment: 983041 }
2022-07-14T21:32:02.889523Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=WindowUpdate { stream_id: StreamId(0), size_increment: 5177345 }
2022-07-14T21:32:02.890245Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=Headers { stream_id: StreamId(1), flags: (0x4: END_HEADERS) }
2022-07-14T21:32:02.890605Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_write: send frame=Data { stream_id: StreamId(1), flags: (0x1: END_STREAM) }
2022-07-14T21:32:02.891366Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Settings { flags: (0x1: ACK) }
2022-07-14T21:32:02.891585Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::proto::settings: received settings ACK; applying Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
2022-07-14T21:32:03.488818Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Headers { stream_id: StreamId(1), flags: (0x4: END_HEADERS) }
2022-07-14T21:32:03.489333Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(1) }
2022-07-14T21:32:03.489526Z DEBUG request{method=POST uri=/ version=HTTP/1.1 otel.kind=server otel.status_code=}:router{query="{\n  topProducts {\n    inStock\n  }\n}" operation_name="" client_name="" client_version="" otel.kind=internal}:execution{otel.kind=internal}:sequence:fetch{otel.kind=internal}:subgraph{name="inventory" otel.kind=internal}:subgraph_request{otel.kind=client net.peer.name=inventory.demo.starstuff.dev net.peer.port=443 http.route=/ net.transport="ip_tcp"}:Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(1), flags: (0x1: END_STREAM) }
2022-07-14T21:32:03.494006Z DEBUG hyper::proto::h1::io: flushed 275 bytes
2022-07-14T21:32:03.494431Z DEBUG hyper::proto::h1::conn: read eof
2022-07-14T21:32:21.760543Z DEBUG hyper::proto::h1::io: parsed 3 headers
2022-07-14T21:32:21.760622Z DEBUG hyper::proto::h1::conn: incoming body is empty
2022-07-14T21:32:21.761226Z DEBUG request{method=GET uri=/plugins/apollo.telemetry/prometheus version=HTTP/1.1 otel.kind=server otel.status_code=}: tower_http::trace::on_request: started processing request
2022-07-14T21:32:21.762107Z DEBUG hyper::proto::h1::io: flushed 171 bytes
2022-07-14T21:32:21.763974Z DEBUG hyper::proto::h1::conn: read eof
</pre>
</details>

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Version 20.04

**Additional context**

### PR DESCRIPTION
**Pull request by Geal**
Fix #1400

the StateMachine was holding a copy of the plugin list to extract the
custom endpoints registered by some of the plugins, like the prometheus
endpoint.

With the Tower refactoring, that plugin list was empty so the endpoints
were not registered.

This commit makes sure that the plugins are only held byt the
RouterServiceFactory instance, and that it implements a method to return
the list of endpoints, that can be used by the StateMachine instance

### CODE REVIEW

File: apollo-router/src/state_machine.rs [hunk 1]
```diff
@@ -85,6 +83,7 @@ impl<S, FA> StateMachine<S, FA>
 where
     S: HttpServerFactory,
     FA: RouterServiceConfigurator + Send,
+    <FA as RouterServiceConfigurator>::RouterServiceFactory: RouterServiceFactory,
```
Comment (garypen) [inline apollo-router/src/state_machine.rs:86]: NIT:

```suggestion
    FA::RouterServiceFactory: RouterServiceFactory,
```

File: apollo-router/src/state_machine.rs [hunk 2]
```diff
@@ -272,23 +267,15 @@ where
...
-            let plugin_handlers: HashMap<String, Handler> = plugins
-                .iter()
-                .filter_map(|(plugin_name, plugin)| {
-                    (plugin_name.starts_with("apollo.") || plugin_name.starts_with("experimental."))
-                        .then(|| plugin.custom_endpoint())
-                        .flatten()
-                        .map(|h| (plugin_name.clone(), h))
-                })
-                .collect();
+            let plugin_handlers: HashMap<String, Handler> = router_factory.custom_endpoints();
```
Comment (garypen) [inline apollo-router/src/state_machine.rs:278]: NIT:
```suggestion
            let plugin_handlers = router_factory.custom_endpoints();
```

File: apollo-router/src/state_machine.rs [hunk 3]
```diff
@@ -342,17 +327,9 @@ where
...
-                    .filter_map(|(plugin_name, plugin)| {
-                        (plugin_name.starts_with("apollo.")
-                            || plugin_name.starts_with("experimental."))
-                        .then(|| plugin.custom_endpoint())
-                        .flatten()
-                        .map(|handler| (plugin_name.clone(), handler))
-                    })
-                    .collect();
+            Ok(new_router_service) => {
+                let plugin_handlers: HashMap<String, Handler> =
```
Comment (garypen) [inline apollo-router/src/state_machine.rs:331]: NIT:
```suggestion
                let plugin_handlers =
```

**Review by garypen — APPROVED** [review]
Looks good. A few small NITS (which may be ignored) to reduce `use` noise and perhaps improve readability.

File: NEXT_CHANGELOG.md [hunk 4]
```diff
@@ -53,6 +53,13 @@ cargo features. We now use reqwest for that.
 
 By [@geal](https://github.com/geal) in https://github.com/apollographql/router/pull/1392
 
+### Expose the custom endpoints from RouterServiceFactory [PR #1402](https://github.com/apollographql/router/pull/1402)
```
Comment (abernix) [inline NEXT_CHANGELOG.md:56]:
```suggestion
### Expose the custom endpoints from RouterServiceFactory ([PR #1402](https://github.com/apollographql/router/pull/1402))
```

File: apollo-router/src/state_machine.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/state_machine.rs:86]: nice

File: NEXT_CHANGELOG.md [hunk 5]
```diff
@@ -53,6 +53,13 @@ cargo features. We now use reqwest for that.
...
+
+Plugin HTTP endpoints registration was broken during the Tower refactoring. We now make sure that the list
+of endpoints is generated from the `RouterServiceFactory` instance.
```
Comment (abernix) [inline NEXT_CHANGELOG.md:59]: :heart: (`` `<3` ``)

**Review by abernix — APPROVED** [review]
LGTM.  One changelog consistency thing.  I have a more meta question about why the test here wasn't failing, but that can be answered elsewhere (the answer is probably _state machines_):

https://github.com/apollographql/router/blob/d5a3a01fe0d8c90a2448dce035cd2f7f052c29bc/apollo-router/src/plugins/telemetry/mod.rs#L1278-L1287

File: apollo-router/src/state_machine.rs [hunk 6]
```diff
@@ -342,17 +327,9 @@ where
...
-                        (plugin_name.starts_with("apollo.")
-                            || plugin_name.starts_with("experimental."))
-                        .then(|| plugin.custom_endpoint())
-                        .flatten()
-                        .map(|handler| (plugin_name.clone(), handler))
-                    })
-                    .collect();
+            Ok(new_router_service) => {
+                let plugin_handlers =
+                    new_router_service.custom_endpoints();
```
Comment (abernix) [inline apollo-router/src/state_machine.rs:332]: Looks like `cargo fmt` wants this in CI.

```suggestion
                let plugin_handlers = new_router_service.custom_endpoints();
```

**Comment by Geal** [comment]
> LGTM. One changelog consistency thing. I have a more meta question about why the test here wasn't failing, but that can be answered elsewhere (the answer is probably state machines):

it was testing the plugin HTTP handlers directly, but here the issue was in how the state machine codes was taking those handlers and setting them up in axum

**Pull request merged**